MetadataExtractor/IO/IndexedCapturingReader.cs

```diff
@@ -51,6 +51,40 @@ public IndexedCapturingReader([NotNull] Stream stream, int chunkLength = Default
 
             _chunkLength = chunkLength;
             _stream = stream ?? throw new ArgumentNullException(nameof(stream));
+
+            try
+            {
+                // For some reason, FileStreams are faster in contiguous mode. Since this is such a a commont case, we
+                // specifically check for it.
+                if (_stream is FileStream)
+                {
+                    _contiguousBufferMode = true;
+                }
+                else
+                {
+                    // If the stream is both seekable and has a length, switch to non-contiguous buffering mode. This
+                    // will use Seek operations to access data that is far beyond the reach of what has been buffered,
+                    // rather than reading the entire file into memory in this case.
+                    _contiguousBufferMode = !(_stream.Length > 0 && _stream.CanSeek);
+                }
+            }
+            catch (NotSupportedException)
+            {
+                // Streams that don't support the Length property have to be handled in contiguous mode.
+                _contiguousBufferMode = true;
+            }
+
+            if (!_contiguousBufferMode)
+            {
+                // If we know the length of the stream ahead of time, we can allocate a Dictionary with enough slots
+                // for all the chunks. We 2X it to try to avoid hash collisions.
+                var chunksCapacity = 2 * (_stream.Length / chunkLength);
+                _chunks = new Dictionary<int, byte[]>((int) chunksCapacity);
```

Comment:
For most data streams we would be unlikely to load all the chunks in the file, so perhaps preallocating enough buckets for the entire file is unnecessary and may even hurt performance by spreading entries across more cache lines than are strictly needed. We'd need to measure performance to be sure one way or another.

Reply:
I only added this after doing a series of performance tests to ensure it would be a win. It's relatively small (~5%) but it was measurable and repeatable.